File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/WcfNonVoidOneWay.cs
using System;
using System.Collections.Generic;
using System.ServiceModel;

namespace Tests.Diagnostics
{
    public class MyAttribute : Attribute
    {
        public bool isOneWay { get; set; }
    }

    [ServiceContract]
    interface IMyService1
    {
        [OperationContract]
        int MyServiceMethod();

        [OperationContract(IsOneWay = true)]
        int MyServiceMethod2(); // Noncompliant {{This method can't return any values because it is marked as one-way operation.}}
//      ^^^

        [OperationContract(IsOneWay = false)]
        int MyServiceMethod3();

        [OperationContract(IsTerminating = true)]
        [My(IsOneWay = true)]
        int MyServiceMethod4();

        [OperationContract(IsOneWay = "mistake")]
        int MyServiceMethod5();

        [OperationContract(IsOneWay = true, AsyncPattern = true)]
        IAsyncResult BeginMyServiceMethod6();
    }
}

File: analyzers/tests/SonarAnalyzer.Test/TestCases/WcfNonVoidOneWay.vb
Imports System
Imports System.Collections.Generic
Imports System.ServiceModel

Namespace Tests.Diagnostics
    Public Class MyAttribute
        Inherits Attribute

        Public Property isOneWay As Boolean
    End Class

    <ServiceContract>
    Interface IMyService1
        <OperationContract>
        Function MyServiceMethod() As Integer

        <OperationContract(IsOneWay:=True)>
        Function MyServiceMethod2() As Integer ' Noncompliant {{This method can't return any values because it is marked as one-way operation.}}
'                                      ^^^^^^^

        <OperationContract(IsOneWay:=False)>
        Function MyServiceMethod3() As String

        <OperationContract(IsTerminating:=True)>
        <My(IsOneWay:=True)>
        Function MyServiceMethod4() As Integer

        ' Error@+1 [BC30934] - Conversion from 'String' to 'Boolean' cannot occur in a constant expression used as an argument to an attribute.
        <OperationContract(Action:="Action", IsOneWay:="mistake")>
        Function MyServiceMethod5() As Integer

        <OperationContract(IsOneWay:=True, AsyncPattern:=True)>
        Function BeginMyServiceMethod6() As IAsyncResult
    End Interface
End Namespace
